migrate cuda implementation of take() from TH to ATen

bdhirsh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45430 migrate cuda implementation of take() from TH to ATen**
* #45283 migrating the take() fn from TH to ATen

Differential Revision: [D24037297](https://our.internmc.facebook.com/intern/diff/D24037297)

bdhirsh committed:
migrate cuda implementation of take() from TH to ATen

[ghstack-poisoned]

bdhirsh committed:
Update on "migrate cuda implementation of take() from TH to ATen"




[ghstack-poisoned]

bdhirsh committed:
Update on "migrate cuda implementation of take() from TH to ATen"




[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/native/cuda/IndexKernel.cu:
```diff
@@ -154,6 +216,54 @@ Tensor & masked_select_out_cuda(Tensor & result, const Tensor & self, const Tens
+  checkAllSameGPU("take", {output_arg, input_arg});
+
+  // TODO: the calls to checkGPU() also assert that all tensors are on CURRENT device, not just the same device as each other (using cudaGetDevice())
+  /*THCAssertSameGPU(THCudaLongTensor_checkGPU(state, 1, index));*/
```
See comment- the original code called `THCudaLongTensor_checkGPU`, which checks that the tensor is on the same device as cudaGetDevice(). It doesn't look like the corresponding ATen function `checkAllSameGPU` performs that check, and I couldn't find ATen code that doesn't something similar.

bdhirsh replied:
I can use that directly. Do you think it would be worth updating `checkAllSameGPU` to do that check too (so that it checks that all passed in tensors are on the same gpu as `current_device()`)? Or would that be too invasive of a change for this PR.

For now I'll just add the extra check directly into the above code.

mrshenli replied:
> Do you think it would be worth updating checkAllSameGPU to do that check too (so that it checks that all passed in tensors are on the same gpu as current_device())?

If doing that, please verify that all callsite of `checkAllSameGPU` expect the tensors are on the current device and also please add comments to `checkAllSameGPU` to highlight its behavior..

bdhirsh committed:
Update on "migrate cuda implementation of take() from TH to ATen"




[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/native/cuda/IndexKernel.cu:
```diff
@@ -5,15 +5,67 @@
 #include <ATen/cuda/CUDAContext.h>
+#include <ATen/cuda/CUDAApplyUtils.cuh>
 #include <ATen/ExpandUtils.h>
+#include <THC/THCTensorInfo.cuh>
```
I included this header to get the MAX_CUTORCH_DIMS and CUTORCH_DIM_WARNING macros from the TH api. Is there a corresponding ATen macro for me to use / better to hardcode them here?

bdhirsh commented on aten/src/ATen/native/cuda/IndexKernel.cu:
```diff
@@ -5,15 +5,67 @@
+  inputInfo.collapseDims();
+  auto numel = input.numel();
+  if (inputInfo.isContiguous()) {
+    cuda::CUDA_tensor_apply2<T, int64_t>(
```
change: using `CUDA_tensor_apply2`, which looks like the corresponding ATen version of `THC_pointwiseApply2`

bdhirsh commented on aten/src/ATen/native/cuda/IndexKernel.cu:
```diff
@@ -5,15 +5,67 @@
+    cuda::CUDA_tensor_apply2<T, int64_t>(
+        output,
+        index,
+        [inputInfo, numel] __device__ (
```
change: Passing in a lambda instead of a `struct TensorTakeOp` object (see the deleted code in `THCTensorIndex.cu`). I assumed that the lack of lambdas was because of the fact that TH is a legacy api.

bdhirsh committed:
Update on "migrate cuda implementation of take() from TH to ATen"




[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/cuda/LegacyTHFunctionsCUDA.cpp:
```diff
@@ -435,144 +435,6 @@ Tensor & _th_index_copy_(Tensor & self, int64_t dim, const Tensor & index, const
-            THCudaShortTensor_take(globalContext().getTHCState(), result_, self_, index_);
-            break;
-        }
-        case ScalarType::Half: {
```
The legacy take() was implemented for `Half` on cuda but not CPU. Should that be consistent?

bdhirsh committed:
Update on "migrate cuda implementation of take() from TH to ATen"




[ghstack-poisoned]

mrshenli reviewed (changes requested):
Adding "request changes" as @bdhirsh mentioned that the new version seems to suffer from a 30% slowdown compared to the existing one.

bdhirsh commented on aten/src/ATen/native/cuda/IndexKernel.cu:
```diff
@@ -154,6 +206,54 @@ Tensor & masked_select_out_cuda(Tensor & result, const Tensor & self, const Tens
+{
+  TORCH_CHECK(output.device().type() == at::kCUDA, "device type of output (", output.device().type(), ") is not on the GPU");
+  TORCH_CHECK(input.device().type() == at::kCUDA, "device type of input (", input.device().type(), ") is not on the GPU");
+  TORCH_CHECK(index.device().type() == at::kCUDA, "device type of index (", index.device().type(), ") is not on the GPUA");
```
gah :(

bdhirsh commented:
> Adding "request changes" as @bdhirsh mentioned that the new version seems to suffer from a 30% slowdown compared to the existing one.

To clarify, the weird slowdown that I saw wasn't from the changes in this PR- separate from this PR, I tried to implement `take` using a call to `gather` (since gather is already efficiently implemented using the TensorIterator pattern), and saw some weird perf numbers.

Some updated benchmark results are below. For context:
1) Each metric is a (median, IQR) pair
2) I ran each benchmark by generated an `input` tensor with 1 million elements, and a corresponding `index` tensor with 1 million elements containing uniform random values on the interval [0,999999]. That seemed like a suitable benchmark to me since it yields a random access pattern on a large tensor. I ran each result for 480 seconds using the `torch.utils._benchmark.Timer` functionality.
3) I implemented take using gather by running the following: `torch.gather(torch.flatten(input), 0, torch.flatten(index)).reshape(index.size())`. I also separately benchmarked a call to `torch.gather(input, 0, index)` to see how much of the time was taken from reshaping the inputs vs. the actual call to gather.

                                                   CPU                              CUDA

Before PR (TH)               (7.80 ms, 0.18ms)              (134.32 us, 0.14 us)

After PR (ATen)              (6.43 ms, 0.05ms)              (134.38 us, 0.03 us)

using `gather`               (4.20 ms, 0.06ms)              (178.56 us, 0.10 us)


A few takeaways:
- This PR by itself isn't a regression- CPU perf has a slight improvement, and CUDA is pretty much unchanged
- `gather` looks like it performs better on CPU, but worse on CUDA. Given that, for now I'm leaning towards leaving the PR in it's current state and not implementing take using gather, but open to other opinions. cc @ngimel

bdhirsh committed:
Update on "migrate cuda implementation of take() from TH to ATen"




[ghstack-poisoned]

bdhirsh committed:
Update on "migrate cuda implementation of take() from TH to ATen"




[ghstack-poisoned]

ngimel commented:
I can't reproduce 30% slowdown with gather, in my benchmarks it is only slightly slower on GPU:
```
import torch
from torch.utils.benchmark import Timer

def gather_fn(inp, ind):
    out = torch.gather(inp.view(-1), 0, ind)
    return out

def take(inp, ind):
    out = inp.take(ind)
    return out

for size in [1000000]:#[100000, 1000000]:
    for false_ratio in [0]:#[0, 0.25, 0.5, 0.75, 1]:
        inp = torch.randn(size, device="cuda", dtype=torch.float32)
        size_ind = int((1 - false_ratio) * size)
        ind = torch.randint(size, (size_ind,), device="cuda")
        timer_take = Timer(stmt="take(inp, ind)", globals=globals())
        timer_gather = Timer(stmt="gather_fn(inp, ind)", globals=globals())
        print(timer_take.blocked_autorange())
        print(timer_gather.blocked_autorange())
```

```
<torch.utils.benchmark.utils.common.Measurement object at 0x7fe804f2b050>
take(inp, ind)
  Median: 87.00 us
  3 measurements, 1000 runs per measurement, 1 thread
<torch.utils.benchmark.utils.common.Measurement object at 0x7fe804f2b050>
gather_fn(inp, ind)
  Median: 90.44 us
  3 measurements, 1000 runs per measurement, 1 thread
```

bdhirsh committed:
Update on "migrate cuda implementation of take() from TH to ATen"




[ghstack-poisoned]

bdhirsh committed:
Update on "migrate cuda implementation of take() from TH to ATen"




[ghstack-poisoned]

bdhirsh committed:
Update on "migrate cuda implementation of take() from TH to ATen"




[ghstack-poisoned]

bdhirsh commented:
> Adding "request changes" as @bdhirsh mentioned that the new version seems to suffer from a 30% slowdown compared to the existing one.

To clarify, the weird slowdown that I saw wasn't from the changes in this PR- separate from this PR, I tried to implement `take` using a call to `gather` (since gather is already efficiently implemented using the TensorIterator pattern), and saw some weird perf numbers.

Some updated benchmark results are below. For context:
1) Each metric is a (median, IQR) pair
2) I ran each benchmark by generated an `input` tensor with 1 million elements, and a corresponding `index` tensor with 1 million elements containing uniform random values on the interval [0,999999]. That seemed like a suitable benchmark to me since it yields a random access pattern on a large tensor. Each result was run using 

                                                   CPU                              CUDA

Before PR                        (7.80 ms, 0.18ms)              (134.32 us, 0.14 us)

After PR                          (6.43 ms, 0.05ms)              (134.38 us, 0.03 us)

using `gather`                (4.20 ms, 0.06ms)              (178.56 us, 0.10 us)


A few takeaways:
- This PR by itself isn't a regression- CPU perf has a slight improvement, and CUDA is pretty much unchanged
- `gather` looks like it performs better on CPU, but worse on CUDA. I also benchmarked running a call to `gather` without any preprocessing (all inputs were 1d contiguous tensors), and got the same result.

> import torch
> from torch.utils.benchmark import Timer
> 
> def gather_fn(inp, ind):
>     out = torch.gather(inp.view(-1), 0, ind)
>     return out
> 
> def take(inp, ind):
>     out = inp.take(ind)
>     return out
> 
> for size in [1000000]:#[100000, 1000000]:
>     for false_ratio in [0]:#[0, 0.25, 0.5, 0.75, 1]:
>         inp = torch.randn(size, device="cuda", dtype=torch.float32)
>         size_ind = int((1 - false_ratio) * size)
>         ind = torch.randint(size, (size_ind,), device="cuda")
>         timer_take = Timer(stmt="take(inp, ind)", globals=globals())
>         timer_gather = Timer(stmt="gather_fn(inp, ind)", globals=globals())
>         print(timer_take.blocked_autorange())
>         print(timer_gather.blocked_autorange())

After a little more digging (and pm's with Natalia), it looks like the perf difference between `take` and `gather` only occurs when the input dtype is an `int64`. From the above benchmarks, the call to `torch.arange` produces `int64` tensors, while `torch.randn` produces `float32` tensors, explaining the benchmark differences.

From what I understand, it sounds like float32 is far and away the common case dtype for tensors, so it's not unreasonable to take the perf hit for int64 and clean up the `take` function to call out to `gather`, simplifying the code quite a bit (and also at some point looking into why `gather` is slower for `int64`s). I can do that in a separate PR- @mrshenli @gchanan thoughts?

bdhirsh committed:
Update on "migrate cuda implementation of take() from TH to ATen"


Differential Revision: [D24037297](https://our.internmc.facebook.com/intern/diff/D24037297)

[ghstack-poisoned]

mrshenli reviewed (approved):
Since this PR is not creating perf regression. I am removing my "request changes" tag. 

Please check with @gchanan regarding if it is OK to take a perf hit for int64 input tensors.

ngimel commented on aten/src/ATen/native/cuda/IndexKernel.cu:
```diff
@@ -154,6 +205,53 @@ Tensor & masked_select_out_cuda(Tensor & result, const Tensor & self, const Tens
+  checkAllSameGPU("take", {output_arg, input_arg, index_arg});
+
+  // additionally check that the above 3 tensors are on the current GPU
+  auto curr_device = at::cuda::current_device();
```
you don't need to do that, code-gen'd deviceGuard is setting current device to the device of the first tensor argument

bdhirsh replied:
thanks for pointing that out

bdhirsh committed:
Update on "migrate cuda implementation of take() from TH to ATen"


Differential Revision: [D24037297](https://our.internmc.facebook.com/intern/diff/D24037297)

[ghstack-poisoned]
